client/network/test/src/service.rs

```diff
@@ -626,6 +626,111 @@ async fn fallback_name_working() {
+	// they can connect.
+	const NEW_PROTOCOL_NAME: &str = "/new-shiny-protocol-that-isnt-PROTOCOL_NAME";
+
+	let listen_addr = config::build_multiaddr![Memory(rand::random::<u64>())];
+	let (node1, mut events_stream1) = TestNetworkBuilder::new()
+		.with_notification_protocol(config::NonDefaultSetConfig {
+			notifications_protocol: NEW_PROTOCOL_NAME.into(),
+			fallback_names: vec![PROTOCOL_NAME.into()],
+			max_notification_size: 1024 * 1024,
+			handshake: None,
+			set_config: Default::default(),
+		})
+		.with_config(config::NetworkConfiguration {
+			listen_addresses: vec![listen_addr.clone()],
+			transport: TransportConfig::MemoryOnly,
+			..config::NetworkConfiguration::new_local()
+		})
+		.build()
+		.start_network();
+
+	let (network2, mut events_stream2) = TestNetworkBuilder::new()
+		.with_set_config(config::SetConfig {
+			reserved_nodes: vec![MultiaddrWithPeerId {
+				multiaddr: listen_addr,
+				peer_id: node1.local_peer_id(),
+			}],
+			..Default::default()
+		})
+		.build()
+		.start_network();
+
+	let receiver = tokio::spawn(async move {
+		let mut old_msg_received = false;
+		let mut new_msg_received = false;
+
+		// Wait for the `NotificationStreamOpened` and send notification over the protocol
+		loop {
+			match events_stream2.next().await.unwrap() {
+				Event::NotificationStreamOpened {
+					remote, protocol, negotiated_fallback, ..
+				} => {
+					assert_eq!(protocol, PROTOCOL_NAME.into());
+					assert_eq!(negotiated_fallback, None);
+					network2.write_notification(
+						remote,
+						protocol,
+						"hello over old name".as_bytes().to_vec(),
+					);
+				},
+				Event::NotificationsReceived { messages, .. } => {
+					assert_eq!(messages.len(), 1);
+					assert_eq!(messages[0].0, PROTOCOL_NAME.into());
+
+					if messages[0].1 == "hello over old name".as_bytes().to_vec() {
+						old_msg_received = true;
+					} else if messages[0].1 == "hello over new name".as_bytes().to_vec() {
+						new_msg_received = true;
+					} else {
+						panic!("invalid message received")
+					}
+
+					if old_msg_received && new_msg_received {
+						break
+					}
+				},
+				_ => {},
+			};
+		}
+	});
+
+	// Wait for the `NotificationStreamOpened` and send notifications over the old and new protocols
+	loop {
+		match events_stream1.next().await.unwrap() {
+			Event::NotificationStreamOpened { remote, protocol, negotiated_fallback, .. }
+				if protocol == NEW_PROTOCOL_NAME.into() =>
+			{
+				assert_eq!(negotiated_fallback, Some(PROTOCOL_NAME.into()));
+				node1.write_notification(
+					remote,
+					protocol,
+					"hello over new name".as_bytes().to_vec(),
+				);
+				node1.write_notification(
+					remote,
+					negotiated_fallback.unwrap(),
+					"hello over old name".as_bytes().to_vec(),
+				);
+			},
+			Event::NotificationsReceived { messages, .. } => {
+				assert_eq!(messages.len(), 1);
+				assert_eq!(messages[0].0, NEW_PROTOCOL_NAME.into());
```

Comment:
@sandreim 

I checked the Polkadot code and it looked like it doesn't check the protocol version of `NotificationsReceived` so notice that it uses the main protocol name for these. Changing this to use legacy protocol would require checking it for every protocol in Substrate and Polkadot which I would rather not do unless it's absolutely necessary as this is one of them temporary hacks again.

Reply:
I think we do the check, at least in my PR https://github.com/paritytech/polkadot/pull/6782/files#diff-64aa50a62fe0bddebd18c842650162a2b1b86b0abda390cc325658bb07d5e222R448 the notifications are handled differently based on the protocol version.

Reply:
But [here](https://github.com/paritytech/polkadot/pull/6782/files#diff-64aa50a62fe0bddebd18c842650162a2b1b86b0abda390cc325658bb07d5e222R351-R353) you're populating `expected_versions` using `PeerId` and it looks like `protocol` (from which `version` is extracted) of `NotificationsReceived` is only used for printing. So if I understood the code correctly and you keep it as is (the protocol version is assigned for `NotificationSteamOpened` and stored for each peer) and you don't use the version of the `NoticationsReceived`, we don't have to change the `NotificationsReceived` handler of every protocol in Substrate/Polkadot.

Reply:
yes, you are right, but the version is stored for each peer on notification stream opened (negotiated version) and is kept by each subsystem until peer disconnects.

Reply:
I remember when initially writing the upgrade code I asked @tomaka for comment and he said the intention was for notifications to be tagged with the "main" protocol ID even if there was a negotiated fallback. So there should be no need to check for legacy protocol names on individual messages, only when establishing the connection.

Reply:
Yes, this fallback system is only relevant for the communication on the wire. When it comes to identifying the protocol through the API, only the main protocol is ever used.

In other words, the main protocol is used both on the wire and as an identifier for that protocol, while the fallbacks are only ever sent on the wire.

Reply:
May be I don't understand something, but if the protocol keeps track of what peers are on what protocol version (after initial connection was established), why can't it use the main protocol name for sending notifications?

Reply:
I haven't tried this yet, I don't think there should be a problem after all ...

Reply:
> May be I don't understand something, but if the protocol keeps track of what peers are on what protocol version (after initial connection was established), why can't it use the main protocol name for sending notifications?

We discussed with @sandreim in DMs and IIRC there was some technical limitation or something that made this less than ideal. But if we can use the main protocol, that'd be better and I can close this PR

Reply:
I think I understood your proposal wrong @altonen, but now what you meant makes sense.